### PR TITLE
Simplify Rectangle and RegularPolygon.

### PR DESCRIPTION
**Pull request by anntzer**
In Rectangle: no need to keep track and update of `self._x1` and
`self._x2` all the time, just when we compute the patch_transform
(which doesn't need to be split out into its own function, or stored
into an instance variable, given that it's never accessed via the
private attribute name).

In RegularPolygon: likewise, we can just compute the transform in
get_patch_transform(), and nowhere else, so the other attributes can
just be plain attributes instead of properties that update the transform
all the time.

## PR Checklist

- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [ ] New features are documented, with examples if plot related.
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [ ] Conforms to Matplotlib style conventions (install `flake8-docstrings` and `pydocstyle<4` and run `flake8 --docstring-convention=all`).
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

File: lib/matplotlib/patches.py [hunk 1]
```diff
@@ -729,59 +729,35 @@ def __init__(self, xy, width, height, angle=0.0, **kwargs):
...
 
     def get_patch_transform(self):
-        self._update_patch_transform()
-        return self._rect_transform
+        # Note: This cannot be called until after this has been added to
+        # an Axes, otherwise unit conversion will fail. This makes it very
+        # important to call the accessor method and not directly access the
+        # transformation member variable.
+        x0, y0, x1, y1 = self._convert_units()
+        bbox = transforms.Bbox.from_extents(x0, y0, x1, y1)
```
Comment (dopplershift) [inline lib/matplotlib/patches.py:758]: These first two lines are identical to `get_bbox()`--refactor to just call that? Would need to get x0/y0 off the bbox.

**Review by dopplershift — COMMENTED** [review]
Is there any normal circumstance where creating these transforms is slow and actually worth caching?

File: lib/matplotlib/patches.py (hunk 1 above)
Comment (anntzer) [reply lib/matplotlib/patches.py:758]: sure

**Comment by anntzer** [comment]
I guess caching may be nice but wouldn't worry about it unless someone shows that this is an actual bottleneck; also in any case this PR doesn't regress on that...

**Comment by QuLogic** [comment]
That's true for `Rectangle`, but not `RegularPolygon`? Also, something that `RegularPolygon` setter could have had, but appears to have been missed, is setting its `stale` property.

**Comment by anntzer** [comment]
RegularPolygon calls `_update_transform` before returning the patch_transform, so we're fine.
OTOH you are right about staleness...

**Comment by QuLogic** [comment]
Yes, but someone could call `get_patch_transform` and just hang onto that even after changing things.

**Review by dopplershift — APPROVED** [review]

**Comment by anntzer** [comment]
Fair point re: RegularPolygon.get_patch_transform; fixed.

**Review by QuLogic — APPROVED** [review]

**Pull request merged**